toke.c: S_intuit_more: Cleanup

khwilliamson opened the pull request:
"This is terrifying code"

At $work there was a story about hardware that collectively handled perhaps billions of phone calls a day.  The interrupt handler was long  spaghetti code which had a single comment /* Subtle */.  Everyone was afraid to touch it, instead creating workarounds for its bugs.  Eventually Management decided to spend the money to rewrite it.

I think this is similar to a portion of S_intuit_more in toke.c, as expressed in #16478.  I have spent some hours going through the code, commenting it and adding white space to make it more legible.  In the process, I discovered some wasted effort, and some missing things.  This PR should make it less terrifying for people to work on

khwilliamson committed:
toke.c: Rmv useless 'else' and outdent

The 'else' functioned only to allow declarations in C89.  Now that we
are on C99, it isn't needed, and gives us more room on the line and
makes it clear there is nothing to follow at the end of the else.

khwilliamson committed:
toke.c: Move some declarations closer to their use

To enhance readability

khwilliamson committed:
toke.c: S_intuit_more: Comments, white-space

I went through this "terrifying" code and analyzed it, adding comments,
and white space to clarify what it is doing

khwilliamson committed:
toke.c: S_intuit_more: Use DeMorgan's laws to clarify

khwilliamson commented:
I did not analyze why the particular weight factors were chosen.  And #16478 indicates this doesn't work as well as claimed

ilmari commented on toke.c:
```diff
@@ -4427,15 +4427,25 @@ S_intuit_more(pTHX_ char *s, char *e)
         return TRUE;
-    if (*s == '-' && s[1] == '>' && (s[2] == '[' || s[2] == '{'))
+
+    if (    *s == '-'
```
I know the code mostly use s `*s` when only looking at one byte, but in places like this where it's looking several bytes ahead using `s[1]` etc., I think it would be clearer to use `s[0]` for the current byte.

khwilliamson replied:
I agree.  I was trying to avoid changing the code unnecessarily.  But now I feel free to

ilmari commented on toke.c:
```diff
@@ -4470,76 +4482,112 @@ S_intuit_more(pTHX_ char *s, char *e)
+
+            /* If it is something like 'a-' or '0-', it is more likely to
+             * be a character class */
+            if (memCHRs("aA01! ", last_un_char))
                 weight += 30;
-            if (memCHRs("zZ79~",s[1]))
+
+            /* If it is something like '-Z' or '-7' (for octal) or '-9' it
+             * is more likely to be a character class */
+            if (memCHRs("zZ79~", s[1]))
```
These comments really should mention the least obvious one of these: `[!-~]` (all printable ASCII characters).

khwilliamson replied:
Thank you, I failed to spot this

ilmari commented on toke.c:
```diff
@@ -4623,6 +4623,11 @@ S_intuit_more(pTHX_ char *s, char *e)
          * one way or another. */
         if (un_char != '\\') {
             seen[un_char]++;
+
+            /* If wrapped around, just leave it at the maximum value */
+            if (seen[un_char] == 0) {
+                seen[un_char]--;
+            }
```
IMO this would be clearer as 
```c
/* Avoid wrapping around */
if (seen[un_char] < 255)
    seen[un_char]++;
```

khwilliamson replied:
Agreed

hvds commented on toke.c:
```diff
@@ -4470,76 +4482,112 @@ S_intuit_more(pTHX_ char *s, char *e)
-    else {
-        /* Otherwise, lean a little more towards it being a character class */
+    else { /* Otherwise, lean a little more towards it being a character class.
+            */
```
Interesting, I much prefer the layout of these comments before the change - when I'm cleaning up gnarly code I'd tend to make exactly the opposite edit (except for the useless blank line at 4474).

khwilliamson replied:
The new PR changes this a bit; and it just seems natural to me to do it similarly in this kind of context

hvds commented on toke.c:
```diff
@@ -4458,9 +4468,11 @@ S_intuit_more(pTHX_ char *s, char *e)
 
     int weight;
+
+    /* Unsigned versions of current character, and the penultimate one */
```
s/penultimate/previous/

I'd expect "the penultimate one" to be the last-but-one character of the string.

khwilliamson replied:
You're right

hvds reviewed (commented):
I'm a bit concerned that this isn't just cleanup, but also makes a variety of functional changes. Can we list all those in one place (perhaps in the ticket), so that we have the chance to more carefully consider which patterns will be affected by them, and thus what sort of breakage this might lead to in the wild?

khwilliamson committed:
toke.c:S_intuit_more: Sort list

This makes it easier to see what's in and what's out.

khwilliamson committed:
toke.c:S_intuit_more: *s -> s[0]

This makes it consistent with other references to 's'

Suggested by ilmari

khwilliamson committed:
toke.c:S_intuit_more: Avoid repeating conditionals

There are two cases that begin with '->'.  Combine them.

khwilliamson commented on toke.c:
```diff
@@ -4470,76 +4482,113 @@ S_intuit_more(pTHX_ char *s, char *e)
+                 * next character after it is a closing bracket, it makes it
+                 * quite likely to be that, and hence a subscript.  If it is
+                 * something else, more mildly a subscript */
+                if (/*{*/ memCHRs("])} =", s[2]))
```
I don't understand why the comment `/*{*/`

hvds replied:
> I don't understand why the comment `/*{*/`

Pretty sure this is to avoid the `}` in the string messing up bracket matching in editors.

khwilliamson replied:
@hvds response is that it is for editors who bounce between matching brackets and don't know to disregard those in string literals, seems quite plausible.  Thanks

khwilliamson commented on toke.c:
```diff
@@ -4493,7 +4493,7 @@ S_intuit_more(pTHX_ char *s, char *e)
     /* Keep track of how many multiple occurrences of the same character there
      * are */
-    char seen[256];
+    U8 seen[256];
```
This change breaks a test in pat.t, showing that the code worked by happenstance.  That test wanted 2000 'b' characters in a row to be treated as a character class.  No one, outside poor generated code, would write the same character multiple times inside a character class: it would reduce to just the first occurrence.  And the code intended to make such things to be interpreted as subscripts, by penalizing it from being a character class with increasing slope for each added redundant character.  But because this array was declared signed, after you got to 127, the next increment would be to -126 or -127 (which I believe is undefined behavior) and that would **subtract** instead of adding additional penalty.  This would continue to -1, then more undefined behavior to get to 0, and the cycle repeats.  The 2000 was arbitrarily chosen; other repeat factors would have resulted in different results

khwilliamson replied:
Actually on a twos complement box, it goes from 127 to -128, which will over the course of many cycles will cause the weight to become more and more negative; exactly the opposite of what the code intends.

khwilliamson replied:
The breakage actually stemmed from a different change, but this remains, and shows the brittleness of the code

khwilliamson commented on toke.c:
```diff
@@ -4446,7 +4447,10 @@ S_intuit_more(pTHX_ char *s, char *e)
     if (s[0] != '{' && s[0] != '[')
         return FALSE;
 
+    /* quit immediately from any errors from now on */
```
I don't know why this is here

khwilliamson replied:
It makes sense to me now.

khwilliamson commented on toke.c:
```diff
@@ -4446,7 +4447,10 @@ S_intuit_more(pTHX_ char *s, char *e)
     if (s[0] != '{' && s[0] != '[')
```
I don't understand this line.  Apparently it is implementing the prolog comment

>  * if we're outside a pattern and it's not { or [, then return FALSE

But we haven't tested yet if we're inside a pattern

khwilliamson replied:
This is just a summary of the function.

khwilliamson commented on toke.c:
```diff
@@ -4428,30 +4428,30 @@ S_intuit_more(pTHX_ char *s, char *e)
 
     /* In a pattern, so maybe we have {n,m}. */
-    if (*s == '{') {
+    if (s[0] == '{') {
```
I don't understand this.  regcurly had very restricted syntax, so the authors could have imagined that anything that matched that would necessarily mean a character class.  But the syntax for regcurly has been relaxed somewhat, and the absence of it being regcurly doesn't mean it can't be a character class

khwilliamson replied:
Hmm, github attached my comment to the wrong line.

khwilliamson replied:
addressed in commentary commit

khwilliamson commented:
Yes the cleanup should be split from any changes

But now that I understand it better, I see how brittle it is, and that some tests pass just because the arbitrarily chosen numbers happened to work.

One thing that I see missing is context.  If this is called immediately when parsing `$a[...]` it would be helpful for this routine to know that there exists a `@a` which would lead the initial bias more towards a subscript

I'd like to know why we aren't getting more reports of failures from the field

khwilliamson committed:
toke.c: S_intuit_more: simplify 1-2 digit case

The way the code worked before this commit when the input is one or two
digits enclosed in [] was:

Originally, the weight is set to 2.  Then if the first byte is a digit,
it gets modified to one of:
 a) If a single digit:     2 - 100 = -98
 b) If precisely 2 digits: 2 - 10  = -8

The code previously then executes the loop with the weight set as above.
It sets un_char to the first digit, and last_un_char to 255, which is
not a \w in ASCII nor is it a \w ASCII equivalent in any of the EBCDIC
character sets ever recognized by Perl.

The loop then executes a switch(); the digit input causes default: to be
taken, none of whose conditionals match, so it subtracts 0 from
the weight (since the character hasn't been seen before, as this is the
first loop iteration).

For the case of the input being a single digit, the loop doesn't execute
again, and the weight is still -98, so it returns TRUE: this is a
subscript.

For the case of the input being precisely two digits, the loop is
executed again, looking at the second digit.  Again default: is chosen
and the weight is:
 a) changed to -8 + 5 = -3 if the second digit is +1 wrt the first
    digit.
 b) changed to -8 - 1 = -7 if the two digits are the same;
 a) unchanged at -8 otherwise

In all cases above, the result is negative, and the function returns
that this is a subscript.

Thus this function always returns TRUE when the input is precisely one
or two digits.

By replacing the current special case by a different special case that
just returns 'TRUE for these inputs, the code is simplified, and opens
the prospect of future simplifications.

This commit moves the 1-2 digit special case to before the weight
declaration, as the weights are now not needed until later; and it  also
prefers pointer subtraction over a memory access.

khwilliamson committed:
S_intuit_more: Add comments; slight simplification

khwilliamson committed:
toke.c:S_intuit_more: Change variable name

'last' is ambiguous.  What is meant here is 'previous'

khwilliamson committed:
toke.c: S_intuit_more: Don't use in-band sentinels

Instead, use a flag to indicate the first time through the loop

The sentinels previously used were unlikely to occur in the input, but
the probability is certainly non-zero; this code may have been written
before Unicode came along.  It's just better to avoid all possible
collisions, and it removes the unclear hack to change a value to 254 in
order to avoid it being the same as the previous value.

khwilliamson committed:
S_intuit_more: Add commentary

Instead of making changes, this comments on where I think such changes
could be profitable.

khwilliamson committed:
S_intuit_more: Allow for EBCDIC

Before comparing EBCDIC code points, one needs to place them in ASCII
order, as not all letters in EBCDIC are contiguous.

khwilliamson commented:
The original pull request is now replaced by other commits which don't change the current behavior

hvds commented on toke.c:
```diff
@@ -4681,6 +4679,7 @@ S_intuit_more(pTHX_ char *s, char *e)
         }   /* End of switch */
 
         seen[un_char]++;
+        first_time = false;
```
This is vulnerable to future addition of a 'continue' further up; an alternative would be to put it in the loop continuation instead:
```
-  for (; s < send; s++) {
+  for (; s < send; s++, first_time = false) {
```

khwilliamson replied:
I never would have thought of this..  Changing

hvds commented on toke.c:
```diff
@@ -4477,11 +4477,20 @@ S_intuit_more(pTHX_ char *s, char *e)
-    /* If begins with '->' ... */
+    /* If begins with '->' ...
+     *
+     * khw: This potentially reads s[1], s[2], and s[3] without checking that
+     * those exist. */
```
Is the string not known to be nul-terminated?

khwilliamson replied:
You're right; I wasn't thinking

hvds commented on toke.c:
```diff
@@ -4609,10 +4632,12 @@ S_intuit_more(pTHX_ char *s, char *e)
             if (s[1]) {
                 if (memCHRs("wds]", s[1]))
                     weight += 100;  /* \w \d \s => strongly charclass */
+                    /* khw: Why not \W \D \S \h \v, etc as well? */
```
Pretty sure at least the negations (`\W \D \S`) were added after this code was last touched.

khwilliamson replied:
My original P.R added them, but you (rightly) objected to it  changing behavior, so blead has never checked for them.  Doing so would be in a new P.R.

hvds commented on toke.c:
```diff
@@ -4625,7 +4650,10 @@ S_intuit_more(pTHX_ char *s, char *e)
+             * character class.
+             *
+             * khw: The rest of the conditionals in this 'case' really should
+             * be subect to an 'else' of this condition */
```
"subject"

khwilliamson replied:
Thanks

hvds commented on toke.c:
```diff
@@ -4685,6 +4718,15 @@ S_intuit_more(pTHX_ char *s, char *e)
             break;
         }   /* End of switch */
 
+        /* khw: 'seen' is declared as a char.  This ++ can cause it to wrap.
+         * This gives different results with compilers for which a plain 'char'
+         * is actually unsigned, versus those where it is signed.  I believe it
+         * is undefined behavior to wrap a 'signed'.  I think it should be
+         * instead declared an unsigned int to make the chances of wrapping
+         * essentially zero.
```
I think making it unsigned would be good. Having it wrap at 8 bits just means the 256'th occurrence adds 0 to the weight, which is no great loss - making them uints doesn't make any difference to the heuristic in any realistic scenario.

hvds reviewed (approved):
My preference would be to avoid trailing comments that are multi-line.
